core/src/peer_id.rs

```diff
@@ -24,13 +24,18 @@ use rand::Rng;
 use std::{convert::TryFrom, fmt, str::FromStr};
 use thiserror::Error;
 
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
 /// Public keys with byte-lengths smaller than `MAX_INLINE_KEY_LENGTH` will be
 /// automatically used as the peer id using an identity multihash.
 const MAX_INLINE_KEY_LENGTH: usize = 42;
 
 /// Identifier of a peer of the network.
 ///
 /// The data is a multihash of the public key of the peer.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```

Comment:
Sorry for only noticing this now but what does `derive(Serialize)` actually result in? I would expect that we serialize to the string representation if the serializer returns true for `is_human_readable` and otherwise returns the efficient byte representation of `Multihash`.

Looking at https://github.com/multiformats/rust-multihash/blob/ebd0d91be8f76ce84000d103243162373d8d08fa/src/multihash.rs#L84-L95, the derive seems to serialize to the internal representation of `Multihash`? This could be argued to be an upstream bug.

Can we add a test to confirm what this serializes to? (And fix the serialization to be _not_ the internal multihash if that is what it turns out to be?)

@mxinden thoughts?

Reply:
Unfortunately I have little to no experience with `serde`. Things that come to my mind:

- We already have a text representation of `PeerId`s. (Actually we have two, see https://github.com/libp2p/rust-libp2p/issues/2259.) Would one implement a custom `Serialize` and `Deserialize` or depend on the internal Rust structures instead?
- In case we do serialize to the internal multihash structure, that would make `multihash` an externally facing dependency of `libp2p`. In other words, every breaking change in `multihash`s serialization code is a breaking change in `libp2p`. Is that something we want?

Reply:
I added a custom `Serialize`/`Deserialize` impl and some unit tests.